File: Cpp/boj_solve/boj14837_3DGeo.cpp
#define _CRT_SECURE_NO_WARNINGS
#include <iostream>
#include <algorithm>
#include <vector>
#include <cmath>
#include <cstring>
#include <cassert>
#include <random>
#include <array>
#include <tuple>
typedef long long ll;
typedef long double ld;
typedef std::pair<int, int> pi;
const ll INF = 1e17;
const int LEN = 1e5 + 1;
const ld TOL = 1e-7;
int N, M, T, Q;
bool zero(const ld& x) { return std::abs(x) < TOL; }
int dcmp(const ld& x) { return std::abs(x) < TOL ? 0 : x > 0 ? 1 : -1; }
int dcmp(const ll& x) { return !x ? 0 : x > 0 ? 1 : -1; }
struct Info { ll area, l, r; };

struct Pos3D {
	ll x, y, z;
	Pos3D(ll X = 0, ll Y = 0, ll Z = 0) : x(X), y(Y), z(Z) {}
	bool operator == (const Pos3D& p) const { return x == p.x && y == p.y && z == p.z; }
	bool operator != (const Pos3D& p) const { return x != p.x || y != p.y || z != p.z; }
	bool operator < (const Pos3D& p) const { return x == p.x ? y == p.y ? z < p.z : y < p.y : x < p.x; }
	ll operator * (const Pos3D& p) const { return x * p.x + y * p.y + z * p.z; }
	Pos3D operator / (const Pos3D& p) const {
		Pos3D ret;
		ret.x = y * p.z - z * p.y;
		ret.y = z * p.x - x * p.z;
		ret.z = x * p.y - y * p.x;
		return ret;
	}
	Pos3D operator + (const Pos3D& p) const { return { x + p.x, y + p.y, z + p.z }; }
	Pos3D& operator += (const Pos3D& p) { x + p.x; y + p.y; z + p.z; return *this; }
	Pos3D operator - (const Pos3D& p) const { return { x - p.x, y - p.y, z - p.z }; }
	Pos3D operator * (const ll& scalar) const { return { x * scalar, y * scalar, z * scalar }; }
	Pos3D& operator *= (const ll& scalar) { x* scalar; y* scalar; z* scalar; return *this; }
	ll Euc() const { return x * x + y * y + z * z; }
	ld mag() const { return sqrtl(Euc()); }
	friend std::istream& operator >> (std::istream& is, Pos3D& p) {
		is >> p.x >> p.y >> p.z;
		return is;
	}
	friend std::ostream& operator << (std::ostream& os, const Pos3D& p) {
		os << p.x << " " << p.y << " " << p.z << "\n";
		return os;
	}
} candi[LEN], willy, MAXP{ INF, INF, INF };
using Face = std::array<int, 3>;
std::vector<Face> Hull3D;
struct Edge {
	int face_num, edge_num;
	Edge(int t = 0, int v = 0) : face_num(t), edge_num(v) {}
};
Pos3D cross(const Pos3D& d1, const Pos3D& d2, const Pos3D& d3) { return (d2 - d1) / (d3 - d2); }
Pos3D cross(const Pos3D& d1, const Pos3D& d2, const Pos3D& d3, const Pos3D& d4) { return (d2 - d1) / (d4 - d3); }
ll dot(const Pos3D& d1, const Pos3D& d2, const Pos3D& d3) { return (d2 - d1) * (d3 - d2); }
ll dot(const Pos3D& d1, const Pos3D& d2, const Pos3D& d3, const Pos3D& d4) { return (d2 - d1) * (d4 - d3); }
bool on_seg_strong(const Pos3D& d1, const Pos3D& d2, const Pos3D& d3) {
	ll ret = dot(d1, d3, d2);
	return !cross(d1, d2, d3).mag() && ret >= 0;
}
bool on_seg_weak(const Pos3D& d1, const Pos3D& d2, const Pos3D& d3) {
	ll ret = dot(d1, d3, d2);
	return !cross(d1, d2, d3).mag() && ret > 0;
}
int ccw(const Pos3D& d1, const Pos3D& d2, const Pos3D& d3, const Pos3D& norm) {
	Pos3D CCW = cross(d1, d2, d3);
	ll ret = CCW * norm;
	return !ret ? 0 : ret > 0 ? 1 : -1;
}
ld area(const std::vector<Pos3D>& H, const Pos3D& norm) {
	ll ret = 0;
	if (H.size() < 3) return ret;
	Pos3D O = H[0];
	int sz = H.size();
	for (int i = 0; i < sz; i++) {
		Pos3D cur = H[i], nxt = H[(i + 1) % sz];
		ret += cross(O, cur, nxt) * norm / norm.Euc();
	}
	return std::abs(ret * .5);
}
bool on_seg_strong(const Pos3D& d1, const Pos3D& d2, const Pos3D& d3) {
	ld ret = dot(d1, d3, d2);
	return zero(cross(d1, d2, d3).mag()) && (ret > 0 || zero(ret));
}
bool on_seg_weak(const Pos3D& d1, const Pos3D& d2, const Pos3D& d3) {
	ld ret = dot(d1, d3, d2);
	return zero(cross(d1, d2, d3).mag()) && ret > 0;
}
//std::vector<Pos3D> graham_scan(std::vector<Pos3D>& C, const Pos3D& norm) {
ld graham_scan(std::vector<Pos3D>& C, const Pos3D& norm) {
	//if (C.size() < 3) {
	//	std::sort(C.begin(), C.end());
	//	return C;
	// }
	if (C.size() < 3) return 0;
	std::vector<Pos3D> H;
	std::swap(C[0], *min_element(C.begin(), C.end()));
	std::sort(C.begin() + 1, C.end(), [&](const Pos3D& p, const Pos3D& q) -> bool {
		int ret = ccw(C[0], p, q, norm);
		if (!ret) return (C[0] - p).Euc() < (C[0] - q).Euc();
		return ret > 0;
		}
	);
	C.erase(unique(C.begin(), C.end()), C.end());
	int sz = C.size();
	for (int i = 0; i < sz; i++) {
		while (H.size() >= 2 && ccw(H[H.size() - 2], H.back(), C[i], norm) <= 0)
			H.pop_back();
		H.push_back(C[i]);
	}
	//return H;
	return area(H, norm);
}
bool collinear(const Pos3D& a, const Pos3D& b, const Pos3D& c) {
	return ((b - a) / (c - b)).Euc() == 0;
}
bool coplanar(const Pos3D& a, const Pos3D& b, const Pos3D& c, const Pos3D& p) {
	return cross(a, b, c) * (p - a) == 0;
}
bool above(const Pos3D& a, const Pos3D& b, const Pos3D& c, const Pos3D& p) {// is p strictly above plane
	return cross(a, b, c) * (p - a) > 0;
}
void prep(std::vector<Pos3D>& p) {//refer to Koosaga'
	shuffle(p.begin(), p.end(), std::mt19937(0x14004));
	int dim = 1;
	for (int i = 1; i < p.size(); i++) {
		if (dim == 1) {
			if (p[0] != p[i]) std::swap(p[1], p[i]), ++dim;
		}
		else if (dim == 2) {
			if (!collinear(p[0], p[1], p[i]))
				std::swap(p[2], p[i]), ++dim;
		}
		else if (dim == 3) {
			if (!coplanar(p[0], p[1], p[2], p[i]))
				std::swap(p[3], p[i]), ++dim;
		}
	}
	assert(dim == 4);
	return;
}
ld dist(const std::vector<Pos3D>& C, const Face& F, const Pos3D& p) {
	Pos3D norm = cross(C[F[0]], C[F[1]], C[F[2]]);
	ll ret = norm * (p - C[F[0]]);
	return std::abs(ret / (norm.mag()));
}
ld get_min_dist(const std::vector<Pos3D>& C, const std::vector<Face>& F, const Pos3D& p) {
	ld MIN = INF;
	for (const Face& face : F) MIN = std::min(MIN, dist(C, face, p));
	return MIN;
}
std::vector<Face> convex_hull_3D(std::vector<Pos3D>& candi) {
	// 3D Convex Hull in O(n log n)
	// Very well tested. Good as long as not all points are coplanar
	// In case of collinear faces, returns arbitrary triangulation
	// Credit: Benq
	// refer to Koosaga'
	prep(candi);
	int sz = candi.size();
	std::vector<Face> faces;
	std::vector<int> active;//whether face is active - face faces outside 
	std::vector<std::vector<int>> vis(sz);//faces visible from each point
	std::vector<std::vector<int>> rvis;//points visible from each face
	std::vector<std::array<Edge, 3>> other;//other face adjacent to each edge of face
	auto ad = [&](const int& a, const int& b, const int& c) -> void {//add face
		faces.push_back({ a, b, c });
		active.push_back(1);
		rvis.emplace_back();
		other.emplace_back();
		return;
		};
	auto visible = [&](const int& a, const int& b) -> void {
		vis[b].push_back(a);
		rvis[a].push_back(b);
		return;
		};
	auto abv = [&](const int& a, const int& b) -> bool {//above
		Face tri = faces[a];
		return above(candi[tri[0]], candi[tri[1]], candi[tri[2]], candi[b]);
		};
	auto edge = [&](const Edge& e) -> pi {
		return { faces[e.face_num][e.edge_num], faces[e.face_num][(e.edge_num + 1) % 3] };
		};
	auto glue = [&](const Edge& a, const Edge& b) -> void {//link two faces by an edge
		pi x = edge(a); assert(edge(b) == pi(x.second, x.first));
		other[a.face_num][a.edge_num] = b;
		other[b.face_num][b.edge_num] = a;
		return;
		};//ensure face 0 is removed when i = 3

	ad(0, 1, 2), ad(0, 2, 1);
	if (abv(1, 3)) std::swap(candi[1], candi[2]);
	for (int i = 0; i < 3; i++) glue({ 0, i }, { 1, 2 - i });
	for (int i = 3; i < N; i++) visible(abv(1, i), i);//coplanar points go in rvis[0]

	std::vector<int> label(sz, -1);
	for (int i = 3; i < sz; i++) {//incremental construction
		std::vector<int> rem;
		for (auto& v : vis[i]) if (active[v]) { active[v] = 0, rem.push_back(v); }
		if (!rem.size()) continue;//hull unchanged

		int st = -1;//start idx
		for (const int& v : rem) {
			for (int j = 0; j < 3; j++) {
				int o = other[v][j].face_num;
				if (active[o]) {//create new face!
					int idx1, idx2;
					std::tie(idx1, idx2) = edge({ v, j });
					ad(idx1, idx2, i);
					st = idx1;
					int cur = rvis.size() - 1;
					label[idx1] = cur;

					std::vector<int> tmp;
					set_union(rvis[v].begin(), rvis[v].end(), rvis[o].begin(), rvis[o].end(), back_inserter(tmp));
					//merge sorted vectors ignoring duplicates

					for (auto& x : tmp) if (abv(cur, x)) visible(cur, x);
					//if no rounding errors then guaranteed that only x > i matters

					glue({ cur, 0 }, other[v][j]);//glue old, new face
				}
			}
		}
		for (int x = st, y; ; x = y) {//glue new faces together
			int X = label[x];
			glue({ X, 1 }, { label[y = faces[X][1]], 2 });
			if (y == st) break;
		}
	}
	std::vector<Face> hull3D;
	for (int i = 0; i < faces.size(); i++) if (active[i]) hull3D.push_back(faces[i]);
	return hull3D;
}

File: Python/boj14837_3DGeo.py
import sys
import math
import random
from functools import cmp_to_key
INPUT = sys.stdin.readline

INF = int(1e17)
TOL = 1e-7


class Pos3D:
    def __init__(self, x: int = 0, y: int = 0, z: int = 0):
        self.x = x
        self.y = y
        self.z = z

    def __eq__(self, other) -> bool:
        return self.x == other.x and self.y == other.y and self.z == other.z

    def __ne__(self, other) -> bool:
        return not self.__eq__(other)

    def __lt__(self, other) -> bool:
        return (self.x, self.y, self.z) < (other.x, other.y, other.z)

    def __add__(self, other) -> 'Pos3D':
        return Pos3D(self.x + other.x, self.y + other.y, self.z + other.z)

    def __sub__(self, other) -> 'Pos3D':
        return Pos3D(self.x - other.x, self.y - other.y, self.z - other.z)

    def __mul__(self, scalar) -> 'Pos3D':
        return Pos3D(self.x * scalar, self.y * scalar, self.z * scalar)

    def __truediv__(self, scalar) -> 'Pos3D':
        return Pos3D(self.x / scalar, self.y / scalar, self.z / scalar)

    def __pow__(self, other) -> int:
        return self.x * other.x + self.y * other.y + self.z * other.z

    def __floordiv__(self, other) -> 'Pos3D':
        return Pos3D(self.y * other.z - self.z * other.y,
                     self.z * other.x - self.x * other.z,
                     self.x * other.y - self.y * other.x)

    def euc(self) -> int:
        return self.x * self.x + self.y * self.y + self.z * self.z

    def mag(self) -> float:
        return math.sqrt(self.euc())

    def __repr__(self) -> str:
        return f"{self.x} {self.y} {self.z}"


O3D = Pos3D(0, 0, 0)


def cross(d1: Pos3D, d2: Pos3D, d3: Pos3D) -> Pos3D:
    return (d2 - d1) // (d3 - d2)


def ccw(d1: Pos3D, d2: Pos3D, d3: Pos3D, nrm: Pos3D):
    torque = cross(d1, d2, d3)
    ret = torque ** nrm
    if not ret:
        return 0
    return 1 if ret > 0 else -1


norm = Pos3D(0, 0, 0)
pivot = Pos3D(0, 0, 0)


# def comp3d(d1: Pos3D, d2: Pos3D) -> bool:
#     f_ = ccw(pivot, d1, d2, norm)
#     if f_ == 0:
#         return (d1 - pivot).euc() < (d2 - pivot).euc()
#     else:
#         return f_ > 0


def comp3d(d1: Pos3D, d2: Pos3D) -> int:
    f_ = ccw(pivot, d1, d2, norm)
    # print("d1", d1.__repr__())
    # print("d2", d2.__repr__())
    # print("pivot", pivot.__repr__())
    # print("norm", norm.__repr__())
    # print(f_)
    if f_ == 0:
        a = (d1 - pivot).euc()
        b = (d2 - pivot).euc()
        if a == b:
            return 0
        elif a < b:
            return -1
        elif a > b:
            return 1
    else:
        return -1 if f_ > 0 else 1


def dot(d1: Pos3D, d2: Pos3D, d3: Pos3D) -> int:
    return (d2 - d1) ** (d3 - d2)


def on_seg_strong(d1: Pos3D, d2: Pos3D, d3: Pos3D) -> bool:
    return cross(d1, d2, d3) == 0 and dot(d1, d3, d2) >= 0


def inner_check(hull: list[Pos3D], nrm: Pos3D, p: Pos3D) -> int:
    sz = len(hull)
    col_ = 0
    for i in range(sz):
        cur = hull[i]
        nxt = hull[(i + 1) % sz]
        if ccw(cur, nxt, p, nrm) < 0:
            return -1
        if not ccw(cur, nxt, p, nrm):
            col_ = 1
    if col_:
        return 0
    return 1


def graham_scan(c: list[Pos3D], nrm: Pos3D, target: Pos3D) -> bool:
    global pivot, norm
    if len(c) < 3:
        c.sort()
        return on_seg_strong(c[0], c[-1], target)

    h = []
    i_ = c.index(min(c))
    c[0], c[i_] = c[i_], c[0]
    pivot, norm = c[0], nrm
    # for p in c:
    #     print(p.x, p.y, p.z, 1)
    c[1:] = sorted(c[1:], key=cmp_to_key(comp3d))
    # c[1:].sort(key=cmp_to_key(comp3d))
    # print(c[0].x, c[0].y, c[0].z)
    # for p in c:
    #     print(p.x, p.y, p.z, 0)

    # sz = len(c)
    # for i in range(sz):
    #     while len(h) >= 2 and ccw(h[-2], h[-1], c[i], nrm) <= 0:
    #         h.pop()
    #     h.append(c[i])
    for p in c:
        while len(h) >= 2 and ccw(h[-2], h[-1], p, nrm) <= 0:
            h.pop()
        h.append(p)
    # for p in h:
    #     print(p.x, p.y, p.z)

    # sz = len(h)
    # for i in range(sz):
    #     cur = h[i]
    #     nxt = h[(i + 1) % sz]
    #     if on_seg_strong(cur, nxt, p):
    #         return True
    # print(norm ** (target - h[0]))
    # print("vec", (target - h[0]).__repr__())
    # print("nrm", norm.__repr__())
    return (norm ** (target - h[0]) == 0) and inner_check(h, nrm, target) > -1


class Face:
    def __init__(self, a, b, c):
        self.v = [a, b, c]

    def norm(self, pos3d: list[Pos3D]) -> Pos3D:
        return cross(pos3d[self.v[0]], pos3d[self.v[1]], pos3d[self.v[2]])

    def visible(self, pos3d: list[Pos3D], i) -> bool:
        return (pos3d[i] - pos3d[self.v[0]]) ** self.norm(pos3d) > 0


# def collinear(a, b, c):
#     return ((b - a) // (c - b)).euc() == 0
#
#
# def coplanar(a, b, c, p):
#     return cross(a, b, c) ** (p - a) == 0
#
#
# def above(a, b, c, p):
#     return cross(a, b, c) ** (p - a) > 0
#
#
# def prep(p):
#     random.shuffle(p)
#     dim = 1
#     assert p[0] != O3D
#     for i in range(1, len(p)):
#         assert p[i] != O3D
#         if dim == 1:
#             if p[0] != p[i]:
#                 p[1], p[i] = p[i], p[1]
#                 dim += 1
#         elif dim == 2:
#             if not collinear(p[0], p[1], p[i]):
#                 p[2], p[i] = p[i], p[2]
#                 dim += 1
#         elif dim == 3:
#             if not coplanar(p[0], p[1], p[2], p[i]):
#                 p[3], p[i] = p[i], p[3]
#                 dim += 1
#     return dim
#
#
# col = True
# cop = True


# def convex_hull_3d(pos3d: list[Pos3D]) -> list[Face]:
#     global col, cop
#     col = False
#     cop = False
#     suf = prep(pos3d)
#     if suf == 2:
#         col = True
#         return []
#     if suf == 3:
#         cop = True
#         return []
#     sz = len(pos3d)
#     vis = [[0] * sz for _ in range(sz)]
#     cur = [Face(0, 1, 2), Face(2, 1, 0)]
#
#     for i in range(3, sz):
#         next_faces = []
#         for f in cur:
#             ret = f.visible(pos3d, i)
#             if not ret:
#                 next_faces.append(f)
#             for k in range(3):
#                 vis[f.v[k]][f.v[(k + 1) % 3]] = ret
#
#         for f in cur:
#             for k in range(3):
#                 a, b = f.v[k], f.v[(k + 1) % 3]
#                 if vis[a][b] != vis[b][a] and vis[a][b]:
#                     next_faces.append(Face(a, b, i))
#
#         cur = next_faces
#
#     return cur
#
#
# def cmp(c: list[Pos3D], face: list[int], p: Pos3D):
#     nrm = cross(c[face[0]], c[face[1]], c[face[2]])
#     ret = nrm ** (p - c[face[0]])
#     return 0 if not ret else (1 if ret > 0 else -1)
#
#
# def inner_check_3d(c: list[Pos3D], faces: list[Face], p: Pos3D):
#     cp = False
#     for face in faces:
#         ab = cmp(c, face.v, p)
#         if ab == 0:
#             cp = True
#         if ab > 0:
#             return -1
#     if cp:
#         return 0
#     return 1


def collinear(a, b, c):
    return ((b - a) // (c - b)).euc() == 0


def coplanar(a, b, c, p):
    return cross(a, b, c) ** (p - a) == 0


def above(a, b, c, p):
    return cross(a, b, c) ** (p - a) > 0


def prep(p):
    random.shuffle(p)
    dim = 1
    assert p[0] != O3D
    for i in range(1, len(p)):
        assert p[i] != O3D
        if dim == 1:
            if p[0] != p[i]:
                p[1], p[i] = p[i], p[1]
                dim += 1
        elif dim == 2:
            if not collinear(p[0], p[1], p[i]):
                p[2], p[i] = p[i], p[2]
                dim += 1
        elif dim == 3:
            if not coplanar(p[0], p[1], p[2], p[i]):
                p[3], p[i] = p[i], p[3]
                dim += 1
    return dim


col = True
cop = True


def convex_hull_3d(candi: list[Pos3D]) -> list[list[int]]:
    global col, cop
    col = False
    cop = False
    suf = prep(candi)
    if suf == 2:
        col = True
        return []
    if suf == 3:
        cop = True
        return []
    sz = len(candi)
    faces = []
    active = []
    vis = [[] for _ in range(sz)]
    rvis = []
    other = []

    def ad(a: int, b: int, c: int) -> None:
        faces.append([a, b, c])
        active.append(True)
        rvis.append([])
        other.append([0, 0, 0])

    def visible(a: int, b: int) -> None:
        vis[b].append(a)
        rvis[a].append(b)

    def abv(a: int, b: int) -> bool:
        tri = faces[a]
        return above(candi[tri[0]], candi[tri[1]], candi[tri[2]], candi[b])

    def edge(e: tuple) -> tuple:
        return faces[e[0]][e[1]], faces[e[0]][(e[1] + 1) % 3]

    def glue(a: tuple, b: tuple) -> None:
        ed = edge(a)
        assert edge(b) == (ed[1], ed[0])
        other[a[0]][a[1]] = b
        other[b[0]][b[1]] = a

    ad(0, 1, 2)
    ad(0, 2, 1)
    if abv(1, 3):
        candi[1], candi[2] = candi[2], candi[1]
    for i in range(3):
        glue((0, i), (1, 2 - i))
    for i in range(3, N):
        visible(abv(1, i), i)
    label = [-1] * sz
    for i in range(3, sz):
        rem = []
        for v in vis[i]:
            if active[v]:
                active[v] = False
                rem.append(v)
        if not rem:
            continue
        st = -1
        for v in rem:
            for j in range(3):
                o = other[v][j][0]
                if active[o]:
                    idx1, idx2 = edge((v, j))
                    ad(idx1, idx2, i)
                    st = idx1
                    cur = len(rvis) - 1
                    label[idx1] = cur
                    tmp = list(set(rvis[v]) | set(rvis[o]))
                    for x in tmp:
                        # print(cur)
                        # print(len(faces))
                        if abv(cur, x):
                            visible(cur, x)
                    glue((cur, 0), other[v][j])
        x = st
        while True:
            idx = label[x]
            y = faces[idx][1]
            glue((idx, 1), (label[y], 2))
            y = faces[idx][1]
            if y == st:
                break
            x = y
    hull3d = [faces[i] for i in range(len(faces)) if active[i]]
    return hull3d


def cmp(c: list[Pos3D], face: list[int], p: Pos3D):
    nrm = cross(c[face[0]], c[face[1]], c[face[2]])
    ret = nrm ** (p - c[face[0]])
    return 0 if not ret else (1 if ret > 0 else -1)


def inner_check_3d(c: list[Pos3D], faces: list[list[int]], p: Pos3D):
    cp = False
    for face in faces:
        ab = cmp(c, face, p)
        if ab == 0:
            cp = True
        if ab > 0:
            return -1
    if cp:
        return 0
    return 1


if __name__ == '__main__':
    # sys.stdin = open('ts2_input.txt', 'r')
    # sys.stdout = open('ts2_output.txt', 'w')
    T = int(INPUT())
    for tc in range(1, T + 1):
        N = int(INPUT())
        C = [Pos3D(*map(int, INPUT().strip().split())) for _ in range(N)]
        Hull3D = convex_hull_3d(C)
        if col:
            C.sort(key=lambda c: (c.x, c.y, c.z))
            f = on_seg_strong(C[0], C[-1], O3D)
            ans = "YES" if f else "NO"
            print(f"Case #{tc}:", ans)
        elif cop:
            norm = cross(C[0], C[1], C[2])
            f = graham_scan(C, norm, O3D)
            ans = "YES" if f else "NO"
            print(f"Case #{tc}:", ans)
        else:
            f = inner_check_3d(C, Hull3D, O3D)
            ans = "YES" if f > -1 else "NO"
            print(f"Case #{tc}:", ans)

